hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/main/java/org/apache/hadoop/mapreduce/v2/app/launcher/ContainerLauncherImpl.java

```diff
@@ -385,6 +385,20 @@ public void run() {
       // TODO: Do it only once per NodeManager.
       ContainerId containerID = event.getContainerID();
 
+      // If the container failed to launch earlier (due to dead node for example),
+      // it has been marked as FAILED and removed from containers during
+      // CONTAINER_REMOTE_LAUNCH event handling.
+      // Skip kill() such container during CONTAINER_REMOTE_CLEANUP as
+      // it is not necessary and could cost 15 minutes delay if the node is dead.
+      if (event.getType() == EventType.CONTAINER_REMOTE_CLEANUP &&
```

Comment:
This should be done in the switch case CONTAINER_REMOTE_CLEANUP

Reply:
I am doing it outside the switch case because in this case I dont want it to go to `getContainer` code which is before the switch case check and `getContainer` will end up giving the a new container and which will evenually fail this check - `!containers.containsKey(containerID)`
```
      Container c = getContainer(event);
```

```
  private Container getContainer(ContainerLauncherEvent event) {
    ContainerId id = event.getContainerID();
    Container c = containers.get(id);
    if(c == null) {
      System.out.println("entering null");
      c = new Container(event.getTaskAttemptID(), event.getContainerID(),
          event.getContainerMgrAddress());
      Container old = containers.putIfAbsent(id, c);
      if(old != null) {
        c = old;
      }
    }
    return c;
  }
```

Reply:
I have made small changes to handle the above case and addressing your comment by moving it inside `switch case CONTAINER_REMOTE_CLEANUP`